### PR TITLE
examples/twr_aloha: refactor

### PR DESCRIPTION
**Pull request by fjmolinas**
### Contribution description

This PR re-works the `twr-aloha` example to make it easier to use as well as include basic radio sleep handling. It also includes a ShellInteration with a basic auto-test that can be used for automatic tests, I plan to add one to the release tests.

### Testing procedure

Run the testing procedure described in the README.

### CODE REVIEW

**Comment by fjmolinas** [comment]
@rdagher might be interested in this one.

File: examples/twr_aloha/README.md [hunk 1]
```diff
@@ -2,67 +2,158 @@
...
+            "site": "saclay",
+            "state": "Alive",
+            "uid": " ",
+            "x": "4.42",
+            "y": "68.3",
+            "z": "6"
+        },
+        ...
+    ]
+}
```
Comment (aabadie) [inline examples/twr_aloha/README.md:67]: Shorter, better:
```suggestion
$ iotlab-experiment  --jmespath="items[*].network_address | sort(@)" get --nodes
[
    dwm1001-2.saclay.iot-lab.info"
]
```

File: examples/twr_aloha/README.md [hunk 2]
```diff
@@ -2,67 +2,158 @@
...
+```
 
-The different algorithm options are:
+Take note of each node's `"network_address"`, when flashing the devices set
+`IOTLAB_NODE` to the value of `"network_address"`, so in this case:
```
Comment (aabadie) [inline examples/twr_aloha/README.md:71]: With the suggestion above, this can be simplified as well.

File: examples/twr_aloha/README.md [hunk 3]
```diff
@@ -2,67 +2,158 @@
...
+$ iotlab-experiment stop
+```
 
-This value can be overridden with a `CFLAG += -DANCHOR_ADDRESS=$(VALUE)`.
+### Setup
 
-Ranging request are sent every 40ms, this value can also be overridden
-by changing the default value of `RANGE_REQUEST_T_US`, e.g:
-`CFLAG += -DRANGE_REQUEST_T_US=10000` to set it at 10ms.
+1. In two separate terminal flash two nodes and open a serial connection to the
```
Comment (aabadie) [inline examples/twr_aloha/README.md:85]:
```suggestion
1. In two separate terminals flash two nodes and open a serial connection to the
```

File: sys/stdio_nimble/Makefile [hunk 4]
```diff
@@ -0,0 +1 @@
+include $(RIOTBASE)/Makefile.base
```
Comment (aabadie) [inline sys/stdio_nimble/Makefile:1]: This file seems unrelated. You added `sys/stdio_nimble` by mistake :)

File: tests/event_ztimer/app.config.test [hunk 5]
```diff
@@ -0,0 +1,9 @@
+# this file enables modules defined in Kconfig. Do not use this file for
```
Comment (aabadie) [inline tests/event_ztimer/app.config.test:1]: Same this change is unrelated :)

**Review by aabadie — COMMENTED** [review]
Found a few minor things but overall the changes looks good. There are unrelated changes added to this PR by mistake !

File: examples/twr_aloha/README.md [hunk 6]
```diff
@@ -2,67 +2,158 @@
...
+    ```shell
+    >twr req C8:0C
+    [twr]: start ranging
+    ```
+
+Both nodes should show the ranging result:
+
+    ```shell
+    > {"t": 55158, "src": 49961, "dst": 51212, "d_cm": 39}
+    ```
```
Comment (rdagher) [inline examples/twr_aloha/README.md:113]:
```suggestion
```shell
> {"t": 55158, "src": 49961, "dst": 51212, "d_cm": 39}
```
```

File: examples/twr_aloha/README.md [hunk 7]
```diff
@@ -2,67 +2,158 @@
...
+    > twr req --help
+    Usage:
+            twr req <short_addr> [-p <proto>] [-c <count>] [-h][-i <ms interval>]
+            - short_addr: short address of request destination
+            - count: number of requests (default: 1)
+            - ms interval: wait interval milliseconds between requests(default: 1000)
+            - proto: twr protocol (ss|ss-ext|ss-ack|ds|ds-ext, default: ss)
+            twr lst on: start listening for ranging requests
+            twr lst off: stop listening for ranging requests
+    ```
```
Comment (rdagher) [inline examples/twr_aloha/README.md:128]:
```suggestion
```shell
> twr req --help
Usage:
        twr req <short_addr> [-p <proto>] [-c <count>] [-h][-i <ms interval>]
        - short_addr: short address of request destination
        - count: number of requests (default: 1)
        - ms interval: wait interval milliseconds between requests(default: 1000)
        - proto: twr protocol (ss|ss-ext|ss-ack|ds|ds-ext, default: ss)
        twr lst on: start listening for ranging requests
        twr lst off: stop listening for ranging requests
```
```

File: examples/twr_aloha/README.md [hunk 8]
```diff
@@ -2,67 +2,158 @@
...
+    {"t": 312545, "src": 51212, "dst": 49961, "d_cm": 35}
+    ```
+
+### Automatic Test
+
+A basic automatic test for the provided shell is included, run with:
+
+    ```shell
+    $ make -C examples/twr_aloha flash test
+    ```
```
Comment (rdagher) [inline examples/twr_aloha/README.md:156]:
```suggestion
```shell
$ make -C examples/twr_aloha flash test
```
```

File: examples/twr_aloha/README.md [hunk 9]
```diff
@@ -2,67 +2,158 @@
...
 
-2. Flash the second node as an anchor
+##### Prerequisites
 
-    $ DW1000_ROLE=0x4 make -C examples/twr_aloha/ flash term
+1. Install iotlab-cli tools
 
-3. On the tag node begin ranging, you will see ranging estimations where
-the "raz" field is "range" in meters.
+    - pip install iotlab-cli
```
Comment (rdagher) [inline examples/twr_aloha/README.md:16]:
```suggestion
    - pip install iotlabcli
```

File: examples/twr_aloha/control.c [hunk 10]
```diff
@@ -74,6 +73,37 @@ static struct uwb_mac_interface _uwb_mac_cbs = (struct uwb_mac_interface){
...
+static void _print_rng_data_cb(void *arg)
+{
+    uwb_core_rng_data_t *rng_data = (uwb_core_rng_data_t *)arg;
+    turo_t ctx;
+    turo_init(&ctx);
+    turo_dict_open(&ctx);
+    turo_dict_key(&ctx, "t");
+    turo_u32(&ctx, rng_data->time);
+    turo_dict_key(&ctx, "src");
+    turo_u32(&ctx, rng_data->src);
```
Comment (rdagher) [inline examples/twr_aloha/control.c:86]: Could be nice to print source addressin hex format using `turo_string` and `sprintf`

File: examples/twr_aloha/control.c [hunk 11]
```diff
@@ -74,6 +73,37 @@ static struct uwb_mac_interface _uwb_mac_cbs = (struct uwb_mac_interface){
...
+    turo_dict_key(&ctx, "dst");
+    turo_u32(&ctx, rng_data->dest);
```
Comment (rdagher) [inline examples/twr_aloha/control.c:88]: Could be nice to print dest address in hex format using `turo_string` and `sprintf`

File: examples/twr_aloha/control.c (hunk 10 above)
Comment (fjmolinas) [reply examples/twr_aloha/control.c:86]: Sure!

**Comment by fjmolinas** [comment]
May I squash?

**Comment by aabadie** [comment]
Please squash!

**Comment by fjmolinas** [comment]
@rdagher you mentioned offline that the test does not work for you. Can you run `make -C examples/twr_aloha/ flash test -j` and log the output here?

**Comment by rdagher** [comment]
> @rdagher you mentioned offline that the test does not work for you. Can you run `make -C examples/twr_aloha/ flash test -j` and log the output here?

Yes I had this problem
```shell
$ make test
EEE
======================================================================
ERROR: test_ifconfig (__main__.TestTWR)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/private/tmp/RIOT/examples/twr_aloha/tests/01-run.py", line 67, in test_ifconfig
    self.shell.parse_netif()
  File "/private/tmp/RIOT/examples/twr_aloha/tests/01-run.py", line 30, in parse_netif
    self._netif = parser.parse(self.ifconfig())
  File "/opt/anaconda3/envs/pepper_riotfp/lib/python3.8/site-packages/riotctrl/shell/__init__.py", line 89, in wrapper
    return func(self, *args, **kwargs)
  File "/private/tmp/RIOT/examples/twr_aloha/twr_shell.py", line 67, in ifconfig
    return self.cmd("ifconfig", timeout, async_)
  File "/opt/anaconda3/envs/pepper_riotfp/lib/python3.8/site-packages/riotctrl/shell/__init__.py", line 101, in cmd
    self._start_replwrap()
  File "/opt/anaconda3/envs/pepper_riotfp/lib/python3.8/site-packages/riotctrl/shell/__init__.py", line 60, in _start_replwrap
    self.riotctrl.term.sendline("")
  File "/opt/anaconda3/envs/pepper_riotfp/lib/python3.8/site-packages/pexpect/pty_spawn.py", line 578, in sendline
    return self.send(s + self.linesep)
  File "/opt/anaconda3/envs/pepper_riotfp/lib/python3.8/site-packages/pexpect/pty_spawn.py", line 569, in send
    return os.write(self.child_fd, b)
OSError: [Errno 5] Input/output error

======================================================================
ERROR: test_listen (__main__.TestTWR)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/private/tmp/RIOT/examples/twr_aloha/tests/01-run.py", line 74, in test_listen
    assert "[twr]: start listening" in self.shell.twr_listen(on=True)
  File "/private/tmp/RIOT/examples/twr_aloha/twr_shell.py", line 54, in twr_listen
    return self.twr_cmd(
  File "/opt/anaconda3/envs/pepper_riotfp/lib/python3.8/site-packages/riotctrl/shell/__init__.py", line 89, in wrapper
    return func(self, *args, **kwargs)
  File "/private/tmp/RIOT/examples/twr_aloha/twr_shell.py", line 51, in twr_cmd
    return self.cmd(cmd, timeout=timeout, async_=False)
  File "/opt/anaconda3/envs/pepper_riotfp/lib/python3.8/site-packages/riotctrl/shell/__init__.py", line 101, in cmd
    self._start_replwrap()
  File "/opt/anaconda3/envs/pepper_riotfp/lib/python3.8/site-packages/riotctrl/shell/__init__.py", line 54, in _start_replwrap
    self.riotctrl.term.read_nonblocking(
  File "/opt/anaconda3/envs/pepper_riotfp/lib/python3.8/site-packages/pexpect/pty_spawn.py", line 460, in read_nonblocking
    incoming = super(spawn, self).read_nonblocking(size)
  File "/opt/anaconda3/envs/pepper_riotfp/lib/python3.8/site-packages/pexpect/spawnbase.py", line 179, in read_nonblocking
    raise EOF('End Of File (EOF). Empty string style platform.')
pexpect.exceptions.EOF: End Of File (EOF). Empty string style platform.

======================================================================
ERROR: test_req (__main__.TestTWR)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/private/tmp/RIOT/examples/twr_aloha/tests/01-run.py", line 78, in test_req
    assert "[twr]: start ranging" in self.shell.twr_req()
  File "/private/tmp/RIOT/examples/twr_aloha/twr_shell.py", line 59, in twr_req
    return self.twr_cmd(
  File "/opt/anaconda3/envs/pepper_riotfp/lib/python3.8/site-packages/riotctrl/shell/__init__.py", line 89, in wrapper
    return func(self, *args, **kwargs)
  File "/private/tmp/RIOT/examples/twr_aloha/twr_shell.py", line 51, in twr_cmd
    return self.cmd(cmd, timeout=timeout, async_=False)
  File "/opt/anaconda3/envs/pepper_riotfp/lib/python3.8/site-packages/riotctrl/shell/__init__.py", line 101, in cmd
    self._start_replwrap()
  File "/opt/anaconda3/envs/pepper_riotfp/lib/python3.8/site-packages/riotctrl/shell/__init__.py", line 54, in _start_replwrap
    self.riotctrl.term.read_nonblocking(
  File "/opt/anaconda3/envs/pepper_riotfp/lib/python3.8/site-packages/pexpect/pty_spawn.py", line 460, in read_nonblocking
    incoming = super(spawn, self).read_nonblocking(size)
  File "/opt/anaconda3/envs/pepper_riotfp/lib/python3.8/site-packages/pexpect/spawnbase.py", line 179, in read_nonblocking
    raise EOF('End Of File (EOF). Empty string style platform.')
pexpect.exceptions.EOF: End Of File (EOF). Empty string style platform.

----------------------------------------------------------------------
Ran 3 tests in 6.780s

FAILED (errors=3)
make: *** [/private/tmp/RIOT/makefiles/tests/tests.inc.mk:22: test] Error 1
```
It was due to the missing dependency on `pyserial` python module, that I solved by doing
```shell
$ pip install pyserial
```

I recommend either adding a requirements.txt file or specifying this in the Prerequisites section of the Readme.md :
```shell
$ pip install iotlabcli riotctrl pyserial
```

So this aside, I tested this PR on both iotlab and locally with two nodes and it works as expected 😎

**Comment by fjmolinas** [comment]
> So this aside, I tested this PR on both iotlab and locally with two nodes and it works as expected

Thanks! I added the suggested addition to the README.

**Comment by fjmolinas** [comment]
@aabadie do you think this one is ok to go?

**Comment by aabadie** [comment]
> do you think this one is ok to go?

It has been reviewed and tested, so yes.

**Review by aabadie — APPROVED** [review]
ACK

**Comment by fjmolinas** [comment]
Thanks for the review and testing @rdagher @aabadie!
